### PR TITLE
Toast remove and timeout

### PR DESCRIPTION
**Pull request by ktabors**
Closes https://jira.corp.adobe.com/browse/RSP-1471

## 📝 Test Instructions:
Storybook toast provider stories have 5 second timeout and close removes them.
Storybook toast provider timeout stories demonstrate different timeout cases.

## 🧢 Your Team:
React Spectrum

Notes - No CSS transitions due to issue with the project changing classnames and the react transition code updates div class names outside of that. This is followup work.

Docs are out of scope and followup work.

Focus being a shift tab is out of scope and followup.

### CODE REVIEW

File: .storybook-v3/custom-addons/provider/index.js [hunk 1]
```diff
@@ -16,7 +16,7 @@ function ProviderUpdater(props) {
   let [localeValue, setLocale] = useState(providerValuesFromUrl.locale || undefined);
   let [themeValue, setTheme] = useState(providerValuesFromUrl.theme || undefined);
   let [scaleValue, setScale] = useState(providerValuesFromUrl.scale || undefined);
-  let [toastPositionValue, setToastPosition] = useState(providerValuesFromUrl.toastPosition || 'top');
+  let [toastPositionValue, setToastPosition] = useState(providerValuesFromUrl.toastPosition || 'bottom');
```
Comment (LFDanLu) [inline .storybook-v3/custom-addons/provider/index.js]: Just curious, why the change from having toast position at the bottom of the storybook instead of at the top?

File: packages/@react-aria/toast/src/useToast.ts [hunk 2]
```diff
@@ -10,17 +10,16 @@
...
 import {PressProps} from '@react-aria/interactions';
-import {ToastProps} from '@react-types/toast';
+import {SpectrumToastProps} from '@react-types/toast';
 import {useId} from '@react-aria/utils';
 import {useMessageFormatter} from '@react-aria/i18n';
 
-interface AriaToastProps extends ToastProps {
-  id?: string
-}
+interface AriaToastProps extends SpectrumToastProps {}
```
Comment (LFDanLu) [inline packages/@react-aria/toast/src/useToast.ts]: I think `AriaToastProps` shouldn't extend `SpectrumToastProps` since this is an aria hook and we want to keep any spectrum specific props out of the prop definition as possible. Perhaps `ToastProps` should be expanded to extend ToastState and any other props that this hook needs?

File: packages/@react-aria/toast/src/useToast.ts [hunk 3]
```diff
@@ -32,8 +31,10 @@ interface ToastAria {
 export function useToast(props: AriaToastProps): ToastAria {
   let {
     id,
+    idKey,
```
Comment (LFDanLu) [inline packages/@react-aria/toast/src/useToast.ts]: small nitpick, feel free to push back
Feel like `idKey` is a bit unintuitive, maybe `toastKey`?

File: packages/@react-aria/toast/test/useToast.test.js [hunk 4]
```diff
@@ -64,10 +65,11 @@ describe('useToast', () => {
   });
 
   it('handles onClose', function () {
-    let {closeButtonProps} = renderToastHook({onClose});
+    let {closeButtonProps} = renderToastHook({onClose, onRemove});
     closeButtonProps.onPress();
 
     expect(onClose).toHaveBeenCalledTimes(1);
+    expect(onRemove).toHaveBeenCalledTimes(1);
```
Comment (LFDanLu) [inline packages/@react-aria/toast/test/useToast.test.js]: Might be worthwhile adding a check that `onRemove` is called with the provided `idKey`

File: packages/@react-spectrum/toast/src/ToastContainer.tsx [hunk 5]
```diff
@@ -14,19 +14,32 @@ import {classNames} from '@react-spectrum/utils';
...
+  let providerProps = useProvider();
+  let toastPlacement = providerProps && providerProps.toastPlacement && providerProps.toastPlacement.split(' ');
+  let containerPosition = toastPlacement && toastPlacement[0];
+  let containerPlacement = toastPlacement && toastPlacement[1];
 
-export function ToastContainer(props: ToastProviderProps): ReactElement {
-  let renderToasts = () => props.toasts.map((toast) =>
-    (<Toast {...toast.props}>{toast.content}</Toast>)
+  let renderToasts = () => toasts.map((toast) =>
+    (<Toast key={toast.props.idKey} {...toast.props} onRemove={onRemove}>{toast.content}</Toast>)
```
Comment (LFDanLu) [inline packages/@react-spectrum/toast/src/ToastContainer.tsx]: Should `{...toast.props}` be put before `key={toast.props.idKey}` to guard against possible overwrites of the key props? Maybe a non-issue since users won't use ToastContainer directly I assume and since toast props shouldn't contain a `key` prop anyways.

File: packages/@react-stately/toast/src/useToastState.ts [hunk 6]
```diff
@@ -10,29 +10,56 @@
...
+  value?: ToastStateValue[]
 }
 
-export function useToastState(props?: ToastStateProps) {
+const TOAST_TIMEOUT = 5000;
+
+export function useToastState(props?: ToastStateProps): ToastState {
   const [toasts, setToasts] = useState(props && props.value || []);
+  const toastsRef = useRef(toasts);
+  toastsRef.current = toasts;
```
Comment (LFDanLu) [inline packages/@react-stately/toast/src/useToastState.ts]: Curious about this ref. Did you have to add it cuz `toasts` wasn't up to date in `onRemove`? Just seems strange that `toastRef` is initialized and then set again to be equal to `toasts`

File: .storybook-v3/custom-addons/provider/index.js (hunk 1 above)
Comment (ktabors) [reply .storybook-v3/custom-addons/provider/index.js]: spectrum design docs say toasts should be on the bottom

File: packages/@react-stately/toast/src/useToastState.ts (hunk 6 above)
Comment (ktabors) [reply packages/@react-stately/toast/src/useToastState.ts]: `onRemove` called by the toast button worked. There was a closure issue with calling `onRemove` for timeouts. Making this a ref solved that closure issue.

File: packages/@react-aria/toast/src/useToast.ts (hunk 2 above)
Comment (ktabors) [reply packages/@react-aria/toast/src/useToast.ts]: I'm looking ad the code and I don't recall why I called this `SpectrumToastProps`, I might have meant `ReactToastPRops` it should probably be renamed. However, sure, I have AriaToastProps extends ToastProps and ToastState with a `toastKey`. I think I extended `SpectrumToastProps` because I was trying to define all the variables like `toastKey` in one place and not spread around the code. I can probably accomplish that a different way.

File: packages/@react-stately/toast/src/useToastState.ts [hunk 7]
```diff
@@ -10,29 +10,56 @@
...
+interface ToastStateProps {
+  value?: ToastStateValue[]
 }
 
-export function useToastState(props?: ToastStateProps) {
+const TOAST_TIMEOUT = 5000;
+
+export function useToastState(props?: ToastStateProps): ToastState {
   const [toasts, setToasts] = useState(props && props.value || []);
+  const toastsRef = useRef(toasts);
```
Comment (snowystinger) [inline packages/@react-stately/toast/src/useToastState.ts]: you may have explained this at some other point, but why are we using a ref?
if it's because state changes after batches, you can use https://dmitripavlutin.com/react-usestate-hook-guide/#updating-the-state-with-a-callback
`setState(prevState => nextState);`

File: packages/@react-stately/toast/src/useToastState.ts [hunk 8]
```diff
@@ -10,29 +10,56 @@
...
+  const onAdd = (content: ReactNode, options: SpectrumToastProps) => {
     let tempToasts = [...toasts];
+    let timeoutId: ReturnType<typeof setTimeout>;
+
+    // set timer to remove toasts
+    if (!(options.actionLabel || options.timeout === 0)) {
+      if (options.timeout < 0) {
+        options.timeout = TOAST_TIMEOUT;
+      }
+      timeoutId = setTimeout(() => onRemove(options.idKey), options.timeout || 5000);
```
Comment (snowystinger) [inline packages/@react-stately/toast/src/useToastState.ts]: should 5000 here also be TOAST_TIMEOUT?

File: packages/@react-stately/toast/test/useToastState.test.js [hunk 9]
```diff
@@ -30,6 +30,8 @@ describe('useToastState', () => {
     let {result} = renderHook(() => useToastState());
     expect(result.current.toasts).toStrictEqual([]);
     act(() => result.current.onAdd(newValue[0].content, newValue[0].props));
+    // timeoutId is dynamic internal so removing it
+    delete result.current.toasts[0].timeoutId;
```
Comment (snowystinger) [inline packages/@react-stately/toast/test/useToastState.test.js]: I don't understand the code comment or why this is being done

Comment (snowystinger) [reply packages/@react-stately/toast/test/useToastState.test.js]: If this is to prevent setTimeouts from firing, you should probably be making use of jest useFakeTimers and should be running them out

Comment (ktabors) [reply packages/@react-stately/toast/test/useToastState.test.js]: the timeout is firing and it adds a dynamic value to the array... I've removing that value

File: packages/@react-stately/toast/src/useToastState.ts (hunk 7 above)
Comment (ktabors) [reply packages/@react-stately/toast/src/useToastState.ts]: Updating the state isn't the issue, it's an initial state within a closure issue. I might have tried that anyway.

File: packages/@react-aria/toast/src/useToast.ts [hunk 10]
```diff
@@ -10,17 +10,16 @@
  * governing permissions and limitations under the License.
  */
 
+import {chain} from '@react-aria/utils';
```
Comment (ktabors) [inline packages/@react-aria/toast/src/useToast.ts]: I forgot to add `onRemove` to an action button when `shouldCloseOnAction`. I just added, but wondered if I should chain `onClose` and `onRemove` and pass that to the action button and close button instead of coding them slightly differently separately.

File: packages/@react-stately/toast/test/useToastState.test.js (hunk 9 above)
Comment (snowystinger) [reply packages/@react-stately/toast/test/useToastState.test.js]: why not let the timeout run in a controlled way? this seems like a test prone to breaking or at least not capable of catching real world use cases since it's doing something that no code we use would ever do. unless it's expected for these to be deleted?

File: packages/@react-aria/toast/src/useToast.ts (hunk 2 above)
Comment (MidnightCoder06) [reply packages/@react-aria/toast/src/useToast.ts]: I think `SpectrumToastProps` is a more appropriate name than `ReactToastProps`
I think it should be `ToastAriaProps` instead of `AriaToastProps`

Comment (ktabors) [reply packages/@react-aria/toast/src/useToast.ts]: I went looking at other aria components and it seems inconsistent. I see `AriaToastProps`, `ToastAriaProps` and `ToastProps` for the props passed into an aria hook.

File: packages/@react-stately/toast/test/useToastState.test.js (hunk 9 above)
Comment (ktabors) [reply packages/@react-stately/toast/test/useToastState.test.js]: I rewrote these tests and next time I run into this issue I won't use `toStrictEqual()` and will use `objectContaining()` to ignore a value.

File: packages/@react-aria/toast/src/useToast.ts [hunk 11]
```diff
@@ -44,25 +47,47 @@ export function useToast(props: AriaToastProps): ToastAria {
...
+
   return {
     toastProps: {
+      ...hoverProps,
       id: useId(id),
       role: 'alert'
     },
     iconProps,
     actionButtonProps: {
+      ...focusProps,
```
Comment (ktabors) [inline packages/@react-aria/toast/src/useToast.ts]: I put the focus on the buttons for now, but it will probably be refactored when we do the focus followup work

File: packages/@react-stately/toast/src/Timer.ts [hunk 12]
```diff
@@ -0,0 +1,20 @@
+export function Timer(callback, delay) {
+  var timerId, start, remaining = delay;
```
Comment (ktabors) [inline packages/@react-stately/toast/src/Timer.ts]: I found this timer online, but I'm not a fan of how it initializes variables. It's pretty much the same as the v2 one.

File: packages/@react-stately/toast/src/Timer.ts [hunk 13]
```diff
@@ -0,0 +1,20 @@
+export function Timer(callback, delay) {
```
Comment (ktabors) [inline packages/@react-stately/toast/src/Timer.ts]: Do I need to add typescript definitions to this?

File: packages/@react-stately/toast/src/timer.ts [hunk 14]
```diff
@@ -0,0 +1,21 @@
+export function Timer(callback, delay) {
```
Comment (ktabors) [inline packages/@react-stately/toast/src/timer.ts]: Do I need to add typescript definitions to this?

File: packages/@react-aria/toast/src/useToast.ts [hunk 15]
```diff
@@ -29,12 +29,15 @@ interface ToastAria {
   closeButtonProps: DOMProps & PressProps
 }
 
-export function useToast(props: AriaToastProps): ToastAria {
+export function useToast(props: ToastAriaProps): ToastAria {
```
Comment (LFDanLu) [inline packages/@react-aria/toast/src/useToast.ts]: I just noticed that the props here contains the state stuff as well. Typically the other aria hooks have state as a separate input parameter for clarity purposes (see https://github.com/adobe-private/react-spectrum-v3/blob/master/packages/%40react-aria/calendar/src/useCalendar.ts#L22 or https://github.com/adobe-private/react-spectrum-v3/blob/master/packages/%40react-aria/calendar/src/useCalendar.ts#L22)

File: packages/@react-aria/toast/src/useToast.ts (hunk 10 above)
Comment (LFDanLu) [reply packages/@react-aria/toast/src/useToast.ts]: I think it is fine as is

File: packages/@react-stately/toast/src/timer.ts (hunk 14 above)
Comment (LFDanLu) [reply packages/@react-stately/toast/src/timer.ts]: eh, might as well. Prob can do it inline with callback being `() => void` maybe and `delay` being a `number`?

File: packages/@react-types/toast/src/index.d.ts [hunk 16]
```diff
@@ -13,22 +13,30 @@
 import {DOMProps, StyleProps} from '@react-types/shared';
 import {ReactNode} from 'react';
 
-export interface ToastOptions {
+export interface ToastOptions extends DOMProps, StyleProps {
```
Comment (dannify) [inline packages/@react-types/toast/src/index.d.ts]: why do the options have dom and style props?

**Comment by dannify** [comment]
This PR will need some follow up changes but merging this for now.

File: packages/@react-types/toast/src/index.d.ts (hunk 16 above)
Comment (ktabors) [reply packages/@react-types/toast/src/index.d.ts]: StyleProps is included for this line of code https://github.com/adobe-private/react-spectrum-v3/blob/master/packages/%40react-spectrum/toast/src/Toast.tsx#L49

DOMProps is included for this line of code https://github.com/adobe-private/react-spectrum-v3/blob/master/packages/%40react-aria/toast/src/useToast.ts#L34
